### ISSUE TITLE
[1.14-beta2] coredns imediatelly crash if Kubernetes API is unavailable

### ISSUE THREAD

**Issue by PierreF** [kubernetes/kubernetes#75414]
**What happened**:

In Kubernetes 1.14 beta2, if kube-apiserver is shortly unavailable (e.g. restart), all CoreDNS pod will crash which cause a short DNS outage

**What you expected to happen**:

Like in Kubernetes 1.13, if kube-apiserver is down, CoreDNS continue to run and provide DNS service.

**How to reproduce it (as minimally and precisely as possible)**:

* Create cluster with kubeadm & calico (any CNI should work).
* Wait for CoreDNS to start
* Restart kube-apiserver
* CoreDNS has crashed (it didn't crashed in Kubernetes 1.13 using the same steps)

```
# POD_CIDR=10.244.0.0/16
# kubeadm init --pod-network-cidr $POD_CIDR --kubernetes-version v1.14.0-beta.2
[...]
# curl https://docs.projectcalico.org/v3.6/getting-started/kubernetes/installation/hosted/kubernetes-datastore/calico-networking/1.7/calico.yaml | sed -e "s?192.168.0.0/16?$POD_CIDR?g" | kubectl apply -f -
# kubectl get po -n kube-system
NAME                                       READY   STATUS    RESTARTS   AGE
[...]
coredns-fb8b8dccf-k8hc2                    1/1     Running   0          26m
coredns-fb8b8dccf-wfbwg                    1/1     Running   0          26m
[...]
# docker rm -f k8s_kube-apiserver_kube-apiserver-node-3_kube-system_1e79449d50c9f3add3dd82d2706ed2f3_0
# kubectl get po -n kube-system
NAME                                       READY   STATUS             RESTARTS   AGE
coredns-fb8b8dccf-k8hc2                    0/1     Running            1          27m
coredns-fb8b8dccf-wfbwg                    0/1     CrashLoopBackOff   1          27m
```

**Anything else we need to know?**:

Log of crashed CoreDNS:

```
# docker logs k8s_coredns_coredns-fb8b8dccf-wfbwg_kube-system_d872299d-4745-11e9-9462-080027c5f494_1
E0315 17:41:50.400737       1 reflector.go:134] github.com/coredns/coredns/plugin/kubernetes/controller.go:317: Failed to list *v1.Endpoints: Get https://10.96.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: connect: connection refused
E0315 17:41:50.400737       1 reflector.go:134] github.com/coredns/coredns/plugin/kubernetes/controller.go:317: Failed to list *v1.Endpoints: Get https://10.96.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: connect: connection refused
log: exiting because of error: log: cannot create log: open /tmp/coredns.coredns-fb8b8dccf-wfbwg.unknownuser.log.ERROR.20190315-174150.1: no such file or directory
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.0-beta.2.66+846a82fecc6959", GitCommit:"846a82fecc69594712040f715d5447bcd445b9c2", GitTreeState:"clean", BuildDate:"2019-03-15T09:35:50Z", GoVersion:"go1.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.0-beta.2", GitCommit:"b1e389e6f7bd798a8dd162f82b918f509ac5291b", GitTreeState:"clean", BuildDate:"2019-03-12T18:01:33Z", GoVersion:"go1.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: Bare-metal (virtualbox)
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`):
```
Linux node-3 4.15.0-46-generic #49-Ubuntu SMP Wed Feb 6 09:33:07 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools: Using kubeadm. Debs file dowloaded from https://console.cloud.google.com/storage/browser/kubernetes-release-dev/bazel/v1.14.0-beta.2.66+846a82fecc6959/
- Others:

```
# docker run --rm -ti k8s.gcr.io/coredns:1.3.1 --version
CoreDNS-1.3.1
linux/amd64, go1.11.4, 6b56a9c
```

/sig network

**Comment by neolit123** [comment]
/sig cluster-lifecycle
@PierreF thanks for the ticket.
cc @johnbelamaric 

on the kubeadm side, CoreDNS was updated from 1.2.6 to 1.3.1 between 1.13 and 1.14.

> Restart kube-apiserver
> CoreDNS has crashed (it didn't crashed in Kubernetes 1.13 using the same steps)

and at this point the CoreDNS pods need a restart too?

**Comment by PierreF** [comment]
> and at this point the CoreDNS pods need a restart too?

To be sure we all talk about the same issue: when kube-apiserver is restarted, coredns crash (so no DNS in whole cluster, which likely cause application outage) and when kube-apiserver is back, CoreDNS are restarted.

My issue is not that DNS never came back, but that with 1.14 CoreDNS no longer continue to run without kube-apiserver and immediately stop running (e.g. they can't tolerate short restart of a single-master).

**Comment by neolit123** [comment]
you try kube-dns instead of CoreDNS with kubeadm:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init-phase/#cmd-phase-addon
so that we can make sure it's a CoreDNS problem.

**Comment by PierreF** [comment]
There is no restart with kube-dns, the DNS service continue to answer while kube-apiserver is down.

```
# POD_CIDR=10.244.0.0/16
# cat > /etc/kubernetes/kubeadm-config.yaml << EOF
apiVersion: kubeadm.k8s.io/v1beta1
kind: ClusterConfiguration
kubernetesVersion: v1.14.0-beta.2
networking:
  podSubnet: "$POD_CIDR"
dns:
  type: "kube-dns"
EOF
# kubeadm init --config /etc/kubernetes/kubeadm-config.yaml
[...]
# curl https://docs.projectcalico.org/v3.6/getting-started/kubernetes/installation/hosted/kubernetes-datastore/calico-networking/1.7/calico.yaml | sed -e "s?192.168.0.0/16?$POD_CIDR?g" | kubectl apply -f -
[...]
# kubectl get po -n kube-system
NAME                                       READY   STATUS    RESTARTS   AGE
[...]
kube-dns-77c76db4b4-wczfp                  3/3     Running   0          3m9s
# docker restart k8s_kube-apiserver_kube-apiserver-node-1_kube-system_d0e0ac765c33aab02537d27fdddd65fa_0
# kubectl get po -n kube-system
NAME                                       READY   STATUS    RESTARTS   AGE
[...
kube-dns-77c76db4b4-wczfp                  3/3     Running   0          4m7s
```

the DNS service continue to respond while kube-apiserver is down:

```
# mv /etc/kubernetes/manifests/kube-apiserver.yaml /tmp  # Force apiserver to stay down
# docker ps | grep kube-apiserver
[ nothing ]
# dig +short kubernetes.default.svc.cluster.local @10.96.0.10
10.96.0.1
```

**Comment by johnbelamaric** [comment]
This is fixed in newer CoreDNS releases: https://github.com/coredns/coredns/issues/2629

The problem was with the K8s client libraries using klog, which crashes if /tmp doesn't exist.

**Comment by johnbelamaric** [comment]
@chrisohaver @rajansandeep can we bump to 1.4.0 in k8s? Or will the Corefile changes be an issue?

**Comment by neolit123** [comment]
@johnbelamaric @chrisohaver @rajansandeep we are separating the 1.14 branch tomorrow EOD PST (code thaw).
so if we want to push this change, it has to happen before that, otherwise we need to backport it to 1.14 at this point.

**Comment by neolit123** [comment]
as per the sig-release slack, SIG Network needs to mention this problem in the "known issues" section https://github.com/kubernetes/kubernetes/issues/74425

**Comment by spiffxp** [comment]
v1.14 release lead here, I am FWIW in favor of you attempting to get this change in prior to v1.14.0, even if it doesn't land prior to code thaw and requires a cherry pick or PR directly to `release-1.14`

**Comment by spiffxp** [comment]
/milestone v1.14
(to keep an eye on this)

**Comment by fturib** [comment]
@johnbelamaric , @neolit123: we did not plan to move to CoreDNS v1.4.0. 
due to deprecation without notice (for k8s user. The notice exist for CoreDNS users)
There is [some bug](https://github.com/coredns/coredns/issues/2654) in v1.4.0, that is fixed but not released. 
I would rather stick with v1.3.1

**Comment by neolit123** [comment]
@fturib thanks, noted.
we rather not bump this last minute anyhow.

can someone from the CoreDNS team post a comment here: https://github.com/kubernetes/kubernetes/issues/74425
about https://github.com/coredns/coredns/issues/2629

as you have better understanding of the problem details.
thank you.

**Comment by johnbelamaric** [comment]
@fturib have you included a relnote for 1.14 regarding the deprecations in the Corefile configuration options? We should get that in for things that were deprecated in 1.3.1 and 1.4.0 now.

**Comment by rajansandeep** [comment]
@johnbelamaric I have included release notes in https://github.com/kubernetes/kubernetes/pull/73610 when CoreDNS was bumped to 1.3.1.
Is there another place we might need to add?

**Comment by johnbelamaric** [comment]
Based on the [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) , it should be an "action required" release note indicating that the deprecation has happened and the functionality will be removed in a future Kubernetes release. Those should be documented with "action required:" in front of them as described in the [release notes guide](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md).

**Comment by neolit123** [comment]
@johnbelamaric 
i'm not sure i understand.
to my understanding, the reason for the coredns crashing problem was a klog related bug: https://github.com/coredns/coredns/pull/2529

why add this under "action required" instead of "known issues" (for 1.3.1)?

**Comment by neolit123** [comment]
@rajansandeep 
> @johnbelamaric I have included release notes in #73610 when CoreDNS was bumped to 1.3.1.
@neolit123 Is there another place we might need to add?

these PR release notes will not be picked by the release note tool at this point, because the editing has already started.

@chrisohaver 's comment here should suffice...
https://github.com/kubernetes/kubernetes/issues/74425#issuecomment-474069189

**Comment by fturib** [comment]
@neolit123 : There are two kinds of information to put in the release note:
1- the "known issue" - which is now in [Chris comment of #74425](https://github.com/kubernetes/kubernetes/issues/74425#issuecomment-474069189)
2- the CoreDNS configuration's deprecation when you move to last version:
   - deprecation tied to v1.3.1 (which is delivered with k8s 1.14) are documented here  in #73610   
   - @johnbelamaric propose to document the warning of `deprecation notice` that goes with v1.4.0 - which are not yet documented anywhere

We would like to inform that some configuration will be deprecated when moving to next release of k8s v1.15, so users have time to prepare their new config of CoreDNS. (CoreDNS is released at a higher speed than k8s, which make difficult to warn on future deprecations before they are effective).

**Comment by neolit123** [comment]
you can propose changes to the release notes directly here:
https://docs.google.com/document/d/1PAbK3D6-GorRf_Fi1U_q-Vvt9wM4MMh-qQfiq-dtOTE/edit

**Comment by jgsqware** [comment]
We are currently facing this issue (with the `klog` issue) and it affect us pretty badly. do you know if there is a way to mitigate it?

**Comment by fturib** [comment]
@jgsqware: I think mitigation will be to use another version of CoreDNS in your deployment. I would prefer @chrisohaver reply/advice on this. (he should be available later today)

**Comment by chrisohaver** [comment]
> do you know if there is a way to mitigate it?

@jgsqware, Upgrade to CoreDNS 1.4.0.

**Comment by chrisohaver** [comment]
@jgsqware, If upgrading to CoreDNS 1.4.0, you _may_ want to disable the reload plugin, as there is a bug in that version that causes CoreDNS to crash when the configuration changes (although, it would crash and get restarted, thus reloading the new config).  If you disable reload, you'll need to restart CoreDNS instances manually to reload config changes.

**Comment by rtheis** [comment]
Can we please get the fix cherry-picked to the next CoreDNS 1.3.x release?

**Comment by chrisohaver** [comment]
FWIW, Historically, the coredns project has maintained a single release track.

**Comment by rtheis** [comment]
@chrisohaver well that certainly presents support problems for Kubernetes' default cluster DNS.  Is the CoreDNS project considering increasing its support window?

**Comment by jgsqware** [comment]
I mitigate the klog issue by mounting an EmptyDir volume to /tmp.

Le mar. 19 mars 2019 à 21:22, Richard Theis <notifications@github.com> a
écrit :

**Comment by chrisohaver** [comment]
> Well that certainly presents support problems for Kubernetes' default cluster DNS.

I agree.  Just letting you know project's history. While the _history_ cannot change, things could change going forward if there is demand.  IMO, the best place to help create demand for this would be to open an issue in CoreDNS.

**Comment by rtheis** [comment]
@chrisohaver thanks for the suggestion.  I've opened https://github.com/coredns/coredns/issues/2708.

**Comment by nikopen** [comment]
As 1.14.0 ships with Coredns 1.3.1, moving this to be properly bumped and resolved in v1.15

/milestone v1.15

@chrisohaver @fturib When the 1.4.0 bug is fixed is it reasonable to ship a 1.14 patch release with the update?

**Comment by chrisohaver** [comment]
The reload bug fix will be included in CoreDNS 1.4.1 release.  But I think the 1.4.1 release also removes an option that is used in the default CoreDNS configuration (which would render it invalid).  So we need to be careful there.  Probably not where we want to go in a patch release.  

We'll see where the discussion in coredns/coredns#2708 goes.  Breaking new ground by releasing a 1.3.2 with the [klog fix](coredns/coredns#2529) may be the best option.

**Comment by fturib** [comment]
@johnbelamaric , @neolit123 : I added a section for CoreDNS in the [section "Deprecation"](https://docs.google.com/document/d/1PAbK3D6-GorRf_Fi1U_q-Vvt9wM4MMh-qQfiq-dtOTE/edit).

The text is:
```
- CoreDNS. The following directives or keywords are deprecated and will be removed in v1.15:
 .  - upstream option of kubernetes plugin, becoming default behavior in v1.15.
    - proxy plugin replaced by forward plugin
```

let me know if further action is needed. Thanks!

**Comment by neolit123** [comment]
@fturib 
SGTM, optionally use quotes around the keywords + uppercase Kubernetes.

**Comment by johnbelamaric** [comment]
yes, looks good to me too

**Comment by kennethredler** [comment]
So [it appears next CoreDNS release will be 1.5.0, not 1.4.1](https://cloud-native.slack.com/archives/C4DF7FP71/p1554149608000900).

What's the most stable choice of CoreDNS version for running 1.14.0 @chrisohaver ❓ 

- [ ] 1.3.1 ❔
- [ ] 1.4.0 ❔ 
- [ ] 1.5.0 ❔ (when it arrives claimedly later "this week")

**Comment by chrisohaver** [comment]
IMO,  1.5.0, barring unforeseen issues, whenever it arrives.

But until then, I think there are three options: 
1. Fall back to 1.3.0.
2. Or use 1.4.0, with the `reload` plugin removed from the Corefile. 
3. Or use 1.3.1 with [workaround](https://github.com/coredns/deployment/pull/138) of mounting an EmptyDir to /tmp.

**Comment by johnbelamaric** [comment]
Yes, I agree 1.5.0 is best, assuming you don't mind tweaking the Corefile.

**Comment by willzhang** [comment]
the question is how to mounting an EmptyDir to /tmp?

**Comment by chrisohaver** [comment]
https://github.com/coredns/deployment/pull/138

On Apr 20, 2019, at 12:26 PM, willgithub <notifications@github.com<mailto:notifications@github.com>> wrote:


the question is how to mounting an EmptyDir to /tmp?

**Comment by dstrebel** [comment]
Hello! We are starting Bug Freeze for 1.15. Is this issue still planned on being incorporated into 1.15?

**Comment by johnbelamaric** [comment]
This is resolved already, there is a workaround and the recommended deployment manifest has been updated to include that.

/close

**Comment by blurpy** [comment]
CoreDNS is still crashing in 1.14.6. Are we not getting a fix for the 1.14-release?

**Comment by johnbelamaric** [comment]
@rajansandeep perhaps we can update the branch?

**Comment by rajansandeep** [comment]
@johnbelamaric if we update the branch, we do it with the deployment workaround?

**Comment by chrisohaver** [comment]
@rajansandeep, yes.  Moving to a version that fixes it is complicated by the required corefile migration.

**Comment by johnbelamaric** [comment]
yes

**Comment by blurpy** [comment]
A workaround for 1.14 sounds good to me!

**Comment by lbiemans** [comment]
My 2ct, in my case iptables was in the way.. Flushed it and coredns started.

**Comment by blurpy** [comment]
Just wondering, since this is fixed in the 1.14-branch now, if 1.14.7 is scheduled for release any time soon? 1.16 is right around the corner, so we are soon to be out of support on 1.13.

**Comment by chrisohaver** [comment]
@blurpy, Kubernetes v1.14.7 cut is planned Wednesday September 18 (tomorrow).

https://groups.google.com/d/msg/kubernetes-dev/Ev8mjKD3h-o/nrJxd3YyAQAJ

**Comment by blurpy** [comment]
Great, thanks!

**Comment by blurpy** [comment]
Hi, I'm back.

We are rolling out 1.14.8, and the backported fix for coredns works nice for us.

However, I'm preparing an upgrade to 1.15.5, and I have the exact same issue again. As soon as the master is unavailable (like during kubeadm upgrade), coredns goes into CrashLoopBackOff.

I diff'ed the deployments of coredns in 1.14.8 and 1.15.5, and the only difference between them is that 1.15.5 is missing the backported volume mount for /tmp to an emptyDir that you added for 1.14.7.

According to @johnbelamaric this should have been fixed in 1.15, but either there is a regression along the way, or it was never fixed. In any case, can this be fixed for the next patch release of 1.15?

**Comment by blurpy** [comment]
I see from the history of the coredns addon that coredns was reverted from 1.5.0 to 1.3.1 without adding the emptyDir hack in 1.15.0-rc1. So it looks like a regression.
